roomserver/api/wrapper.go

```diff
@@ -47,34 +47,66 @@ func SendEventWithState(
 	ctx context.Context, rsAPI RoomserverInternalAPI, state *gomatrixserverlib.RespState,
 	event gomatrixserverlib.HeaderedEvent, haveEventIDs map[string]bool,
 ) error {
-	outliers, err := state.Events()
+	isCurrentState := map[string]struct{}{}
+	for _, se := range state.StateEvents {
+		isCurrentState[se.EventID()] = struct{}{}
+	}
+
+	authAndStateEvents, err := state.Events()
 	if err != nil {
 		return err
 	}
 
 	var ires []InputRoomEvent
-	for _, outlier := range outliers {
-		if haveEventIDs[outlier.EventID()] {
+	var stateIDs []string
+
+	for _, authOrStateEvent := range authAndStateEvents {
+		if authOrStateEvent.StateKey() == nil {
 			continue
 		}
-		ires = append(ires, InputRoomEvent{
-			Kind:         KindOutlier,
-			Event:        outlier.Headered(event.RoomVersion),
-			AuthEventIDs: outlier.AuthEventIDs(),
-		})
-	}
+		if haveEventIDs[authOrStateEvent.EventID()] {
+			continue
+		}
+
+		// Work out whether we should store the event as an outlier
+		// or if we should send it as a rewrite event.
+		storeAsOutlier := false
+		if authOrStateEvent.Type() == event.Type() && *authOrStateEvent.StateKey() == *event.StateKey() {
```

Comment:
Won't this panic if `event` isn't a state event?

Reply:
There's a guard before it:
```
if authOrStateEvent.StateKey() == nil {
    continue
}
```

Reply:
There's no guard for `event` though which is also being de-referenced.

Reply:
This is still an issue.